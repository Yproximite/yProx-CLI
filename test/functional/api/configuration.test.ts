import { readFileSync } from 'fs';
import { createFakeEnv } from '../fake-env';

const packageJson = readFileSync(`${__dirname}/../fixtures/modern-project/package.json`, 'utf8');
const yproxCliConfigJs = readFileSync(`${__dirname}/../fixtures/modern-project/yprox-cli.config.js`, 'utf8');

describe('api: configuration', () => {
  beforeEach(() => {
    // @ts-ignore
    process.exit = jest.fn();
    console.error = jest.fn();
  });

  afterEach(() => {
    // @ts-ignore
    process.exit.mockRestore();
    // @ts-ignore
    console.error.mockRestore();
  });

  it('should load conf from `yprox-cli.config.js`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
  });

  it('should load conf from `package.json`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
  });

  it('should throw an error when having both config from `package.json` and `yprox-cli.config.js`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(process.exit).toHaveBeenCalledWith(1);
    expect(console.error).toHaveBeenNthCalledWith(1, '[08:30:00] error :: Your configuration is invalid.');
    expect(console.error).toHaveBeenNthCalledWith(
      2,
      "[08:30:00] error :: You can't configure yprox-cli with yprox-cli.config.js and package.json at the same time."
    );

    await cleanup();
  });

  it('with YPROX_CLI_IGNORE_CONFIG_FILE=true, should load config from `package.json` even if we have two config files', async () => {
    process.env.YPROX_CLI_IGNORE_CONFIG_FILE = 'true';

    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
    delete process.env.YPROX_CLI_IGNORE_CONFIG_FILE;
  });

  it('with YPROX_CLI_IGNORE_PACKAGE_JSON_FILE=true, should load config from `yprox-cli.config.js` even if we have two config files', async () => {
    process.env.YPROX_CLI_IGNORE_PACKAGE_JSON_FILE = 'true';

    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
    delete process.env.YPROX_CLI_IGNORE_PACKAGE_JSON_FILE;
  });
});
