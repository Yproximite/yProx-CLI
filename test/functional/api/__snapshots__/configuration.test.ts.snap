// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api: configuration should load conf from "yprox-cli.config.js" 1`] = `
Object {
  "autoprefixer": Object {},
  "babel": true,
  "cssnano": Object {
    "preset": "default",
  },
  "eslint": Object {
    "extensions": Array [
      ".js",
      ".vue",
    ],
  },
  "gifsicle": Object {
    "interlaced": true,
  },
  "handlers": Object {
    "javascript": Object {
      "babel": Object {
        "exclude": Array [
          "/node_modules/**",
        ],
      },
      "commonjs": Object {},
      "json": Object {},
      "nodeResolve": Object {
        "browser": true,
        "extensions": Array [
          ".mjs",
          ".js",
          ".json",
          ".node",
          ".vue",
        ],
        "mainFields": Array [
          "module",
        ],
      },
      "shims": Object {},
      "vue": Object {},
    },
    "sass": Object {},
  },
  "jpegtran": Object {
    "progressive": true,
  },
  "optipng": Object {
    "optimizationLevel": 5,
  },
  "path": Object {},
  "svgo": Object {
    "plugins": Array [
      Object {
        "removeViewBox": true,
      },
    ],
  },
  "terser": Object {},
}
`;
